File: packages/coterie-js/src/core/llm/scripted.ts
/** Test / demo double. Replays a queue of pre-recorded responses. */

import { readFileSync } from "node:fs";

import type { LLMClient, LLMMessage } from "./base.js";

export class ScriptedLLMError extends Error {}

export class ScriptedLLMClient implements LLMClient {
  private responses: string[];
  readonly calls: { system: string; messages: LLMMessage[] }[] = [];

  constructor(responses: Iterable<string>) {
    this.responses = [...responses];
  }

  static fromFixture(path: string): ScriptedLLMClient {
    const data = JSON.parse(readFileSync(path, "utf8"));
    if (!Array.isArray(data) || !data.every((x) => typeof x === "string")) {
      throw new Error(`fixture at ${path} must be a JSON array of strings`);
    }
    return new ScriptedLLMClient(data);
  }

  queue(response: string): void {
    this.responses.push(response);
  }

  async chat(system: string, messages: LLMMessage[]): Promise<string> {
    this.calls.push({ system, messages });
    if (this.responses.length === 0) {
      throw new ScriptedLLMError(
        `ScriptedLLMClient exhausted after ${this.calls.length} call(s); last system: ${system.slice(0, 100)}`,
      );
    }
    return this.responses.shift()!;
  }
}

File: packages/coterie-py/src/coterie/core/llm/scripted.py
"""Test / demo doubles for LLMClient.

Slide 20: protocol-typed deps make fakes trivial. Slide 21: ScriptedLLMClient
ships in the real package so unit tests and offline demos use the same
implementation.
"""

import json
from collections import deque
from pathlib import Path
from typing import Iterable

from coterie.core.llm.base import LLMClient


class ScriptedLLMError(RuntimeError):
    """Raised when a ScriptedLLMClient is called with no remaining responses."""


class ScriptedLLMClient(LLMClient):
    """Replays a queue of pre-recorded responses.

    Two use cases (slide 21):
    1. Unit / integration tests — no API key, no network, deterministic.
    2. Offline demo on stage — agents still run for real, only the LLM round-trip is replayed.
    """

    def __init__(self, responses: Iterable[str]) -> None:
        self._responses: deque[str] = deque(responses)
        self.calls: list[dict] = []

    @classmethod
    def from_fixture(cls, path: str | Path) -> "ScriptedLLMClient":
        data = json.loads(Path(path).read_text())
        if not isinstance(data, list) or not all(isinstance(x, str) for x in data):
            raise ValueError(f"fixture at {path} must be a JSON array of strings")
        return cls(data)

    def queue(self, response: str) -> None:
        self._responses.append(response)

    def chat(self, system: str, messages: list[dict]) -> str:
        self.calls.append({"system": system, "messages": messages})
        if not self._responses:
            raise ScriptedLLMError(
                f"ScriptedLLMClient exhausted after {len(self.calls)} call(s); "
                f"last system prompt: {system[:100]!r}"
            )
        return self._responses.popleft()
